### DISCUSSION TITLE
"models, how do they work?"

### DISCUSSION THREAD

**Discussion by metazool** [ukceh-rse/community#7] [Social Coding]
### I'd like to work on...

I've never worked on a modelling project. People talk about models as if they're sure about exactly what they are. There's a lot of model development and I'd like to understand at a really "models 101" level how they are meant to work, what frameworks and common patterns there are for implementing one. How simply could one make a toy one? Are models even _really a thing_, at an ontological level? I have so many basic questions 

### Collaboration

I am equally happy working as a group or independently.

### My aim is to...

Implement a toy model from scratch to understand what is involved!

**Comment by mattjbr123** [comment]
Definitely interested in this, having done some modelling in my time, I can probably help, a bit? Or at least be willing to discuss!

**Comment by samharrison7** [comment]
I love this question! :)

Here's my take: There's no real definition of what a "model" is. In my mind, it is anything that is intended to be some kind of (potentially very abstract) representation of a physical entity, which takes some input and transforms it into some output. It could be anything from a really simple function (like $E = mc^2$) right up to complex earth system models with 100,000s of lines of code. They could be process-based - where they try to mimic physical systems by linking together algorithms for specific physical processes - or more data-driven/statistical, such as machine learning models that can give good predictions of a physical system without knowing anything about the physical processes that make up that system. Models don't necessarily have to be written in code, they could be implemented in something like a spreadsheet instead.

Different disciplines will probably have different frameworks, languages and common patterns, but there is definitely no one-size-fits-all when it comes to modelling in general. Generally, a "model" will be a big lump of code (maybe nicely split into sensible files, modules, classes, etc). But more and more we're thinking about how we could use different design patterns to offer more flexibility, such as using component-based architecture to split models up into more flexible, potentially interchangeable components.

P.S. Hope you don't mind me as a non-RSE interjecting here?!

**Comment by jmarshrossney** [comment]
Hey Sam - we absolutely **love** a non-RSE interjecting! The idea was to eventually publicise this discussion forum (and probably move it to NERC-CEH), but we wanted to have something already there before we did that.

I meant to follow this up in light of the excellent Euro-CSDMS modelling workshop (Note to self: when that forum launches we should include a link to it in the pinned post).

At that workshop we discussed the idea of crowd-sourcing a "toybox" of (toy) models, all adopting the [Basic Model Interface](https://bmi.readthedocs.io/en/stable/). This would be mainly for pedagogical purposes, but I also think there is enormous merit to routinely comparing results from complicated models against simpler ones. Anyway, it's an awesome idea and I'm genuinely really excited at the thought of building some toy models with a purpose.

So, Jo, would you be interested in turning this whole exercise into a contribution to the toybox? This doesn't really introduce any more complexity, it just means there's an extra step at the end where we 'BMI-ify' the model (which just means subclassing a Python ABC and implementing a few of its methods).

**Reply by metazool** [reply]
I'm belatedly catching up with notifications and yes! This seems like a nicely sized self-contained exercise. I will aim for it as a learning and documentation project in the quiet days before Christmas (Advent of Models?).

Thanks for contributing both reflections and a lot of relevant examples @samharrison7 - reaching out to folks who aren't explicitly in RSE roles but have a lot of their behaviours is a good result for this forum :)

**Reply by samharrison7** [reply]
I like the idea of Advent of Models - one new model per day?!

**Reply by metazool** [reply]
i've been doing Tiny Code Christmas on the tic80 again (intro mode last year, extra mode this year...) Implementing one or more of the "simplest-in-class" models that @jmarshrossney suggested below, in Lua on a fantasy console, would be a nice reason to keep going after the challenges end :)

https://tcc.lovebyte.party

Tiny Model Christmas next year, maybe?

**Comment by jmarshrossney** [comment]
Some suggestions for kind of the simplest-in-class models:

## Evolving a system forwards in continuous time

- Simple energy balance models (e.g. for predicting temperature of the Earth)
- Lotka-Volterra (simple predator-prey) model
- Solving some simple chemical system in a box (useful for me to better understand EMEP)
- Bucket hydrological model
- Very minimal land-surface model (useful for me to understand JULES)

## Evolving a system in both space and time

- Double pendulum
- Solar system
- 2d diffusion (advection is more complicated)
- 2d random walk (which is actually the same as above)
- Vicsek model (minimal model for collective motion, e.g. flocking birds)

## Statistical mechanical models

(More about sampling a system using Markov Chain Monte Carlo, as opposed to evolving the system in physical time or space.)

- Ising model (binary variables on a 2d lattice)

## Agent-based models

I've never looked at these, but they come up in population ecology among other places. I don't have any minimal examples.

## Evolving a system forward in discrete time

- Conway's game of life
- Generating fractals

I might have gotten carried away. I think the models in the first and second sections would be most appropriate to start.

**Reply by jmarshrossney** [reply]
I would say that the double pendulum, or something that evolves a system in one time and one spatial dimension, is probably the most pedagogical, because you can 'see' the evolution.

Although it's also nice to get intuition about how systems equilibriate, so the first section.

**Reply by samharrison7** [reply]
Here's a BMI-ified version of the leaky bucket hydrological model! https://github.com/eWaterCycle/leakybucket-bmi

**Comment by samharrison7** [comment]
That's a super list of models!

I really like the idea of a toybox of models, particularly if they all expose a BMI (though maybe a BMI isn't appropriate for all models - that's a discussion point in it's own right). This also reminds me that the BMI examples use the [heat equation model](https://en.wikipedia.org/wiki/Heat_equation), which in itself is a nice simple toy model to play with. For example, [here's the Python version](https://github.com/csdms/bmi-example-python). And (because everyone needs a bit more Fortran in their lives) [here's the Fortran version](https://github.com/csdms/bmi-example-fortran).

**Comment by metazool** [comment]
https://scipy-cookbook.readthedocs.io/items/LoktaVolterraTutorial.html - classically prey-predator, but also plant-pollinator applications, multispecies, mutualism
